Inherent associated type projections are not normalized after substitution

I tried this code:

```rust
#![feature(inherent_associated_types)]
#![allow(incomplete_features)]

struct S<T>(T);
impl<T: O> S<T> {
    type P = <T as O>::P;
}
trait O {
    type P;
}
impl O for i32 {
    type P = String;
}
fn main() {
    let _: S<i32>::P = String::new();
}
```

I expected to see it compile successfully.
Instead, it was rejected with the following error message:

```
error[E0308]: mismatched types
  --> p0.rs:15:24
   |
15 |     let _: S<i32>::P = String::new();
   |            ---------   ^^^^^^^^^^^^^ expected associated type, found struct `String`
   |            |
   |            expected due to this
   |
   = note: expected associated type `<i32 as O>::P`
                       found struct `String`
   = help: consider constraining the associated type `<i32 as O>::P` to `String` or calling a method that returns `<i32 as O>::P`
   = note: for more information, visit https://doc.rust-lang.org/book/ch19-03-advanced-traits.html
help: call `Into::into` on this expression to convert `String` into `<i32 as O>::P`
   |
15 |     let _: S<i32>::P = String::new().into();
   |                                     +++++++
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc -Vv`:
```
rustc 1.67.0-nightly (53e4b9dd7 2022-12-04)
binary: rustc
commit-hash: 53e4b9dd74c29cc9308b8d0f10facac70bb101a7
commit-date: 2022-12-04
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

@rustbot label F-inherent_associated_types
@rustbot claim

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"fmease"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_hir_analysis/src/astconv/mod.rs
+++ b/compiler/rustc_hir_analysis/src/astconv/mod.rs
@@ -1930,6 +1930,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                     adt_substs,
                 );
                 let ty = tcx.bound_type_of(assoc_ty_did).subst(tcx, item_substs);
+                let ty = self.normalize_ty(span, ty);
                 return Ok((ty, DefKind::AssocTy, assoc_ty_did));
             }
         }

--- a/src/test/ui/associated-inherent-types/normalize-projection-0.rs
+++ b/src/test/ui/associated-inherent-types/normalize-projection-0.rs
@@ -1,0 +1,22 @@
+// check-pass
+
+#![feature(inherent_associated_types)]
+#![allow(incomplete_features)]
+
+struct S<T>(T);
+
+impl<T: O> S<T> {
+    type P = <T as O>::P;
+}
+
+trait O {
+    type P;
+}
+
+impl O for i32 {
+    type P = String;
+}
+
+fn main() {
+    let _: S<i32>::P = String::new();
+}

--- a/src/test/ui/associated-inherent-types/normalize-projection-1.rs
+++ b/src/test/ui/associated-inherent-types/normalize-projection-1.rs
@@ -1,0 +1,22 @@
+// check-pass
+
+#![feature(inherent_associated_types)]
+#![allow(incomplete_features)]
+
+struct S;
+
+impl S {
+    type P<T: O> = <T as O>::P;
+}
+
+trait O {
+    type P;
+}
+
+impl O for i32 {
+    type P = String;
+}
+
+fn main() {
+    let _: S::P<i32> = String::new();
+}

